packages/typespec-vscode/src/utils.ts

```diff
@@ -102,3 +93,211 @@ export async function loadModule(
+    logger.debug(`Failed to fetch from url: ${url}`, [e]);
+    return undefined;
+  }
+}
+
+export interface ExecOutput {
+  stdout: string;
+  stderr: string;
+  exitCode: number;
+  error: any;
+  spawnOptions: SpawnOptions;
+}
+export interface spawnExecutionEvents {
+  onStdioOut?: (data: string) => void;
+  onStdioError?: (error: string) => void;
+  onError?: (error: any, stdout: string, stderr: string) => void;
+  onExit?: (code: number | null, stdout: string, stderror: string) => void;
+}
+
+/**
+ * The promise will be rejected if the process exits with non-zero code or error occurs. Please make sure the rejection is handled property with try-catch
+ *
+ * @param exe
+ * @param args
+ * @param cwd
+ * @returns
+ */
+export function spawnExecutionAndLogToOutput(
+  exe: string,
+  args: string[],
+  cwd: string,
+): Promise<ExecOutput> {
+  return spawnExecution(exe, args, cwd, {
+    onStdioOut: (data) => {
+      logger.info(data.trim());
+    },
+    onStdioError: (error) => {
+      logger.error(error.trim());
+    },
+    onError: (error) => {
+      if (error?.code === "ENOENT") {
+        logger.error(`Cannot find ${exe} executable. Make sure it can be found in your path.`);
+      }
+    },
+  });
+}
+
+/**
+ * The promise will be rejected if the process exits with non-zero code or error occurs. Please make sure the rejection is handled property with try-catch
+ *
+ * @param exe
+ * @param args
+ * @param cwd
+ * @param on
+ * @returns
+ */
+export function spawnExecution(
+  exe: string,
+  args: string[],
+  cwd: string,
+  on?: spawnExecutionEvents,
+): Promise<ExecOutput> {
+  const shell = process.platform === "win32";
+  const cmd = shell && exe.includes(" ") ? `"${exe}"` : exe;
+  let stdout = "";
+  let stderr = "";
+
+  const options: SpawnOptions = {
+    shell,
+    stdio: "pipe",
+    windowsHide: true,
+    cwd,
+  };
+  const child = spawn(cmd, args, options);
+
+  child.stdout!.on("data", (data) => {
+    stdout += data.toString();
+    if (on && on.onStdioOut) {
+      try {
+        on.onStdioOut!(data.toString());
+      } catch (e) {
+        logger.error("Unexpected error in onStdioOut", [e]);
+      }
+    }
+  });
+  child.stderr!.on("data", (data) => {
+    stderr += data.toString();
+    if (on && on.onStdioError) {
+      try {
+        on.onStdioError!(data.toString());
+      } catch (e) {
+        logger.error("Unexpected error in onStdioError", [e]);
+      }
+    }
+  });
+  if (on && on.onError) {
+    child.on("error", (error: any) => {
+      try {
+        on.onError!(error, stdout, stderr);
+      } catch (e) {
+        logger.error("Unexpected error in onError", [e]);
+      }
+    });
+  }
+  if (on && on.onExit) {
+    child.on("exit", (code) => {
+      try {
+        on.onExit!(code, stdout, stderr);
+      } catch (e) {
+        logger.error("Unexpected error in onExit", [e]);
+      }
+    });
+  }
+  return new Promise((res, rej) => {
+    child.on("error", (error: any) => {
+      rej({
```

Comment:
Can we not reject( in that way it will throw exception), just resolve? we will check the retcode to verify if the execution succeed or not.

Reply:
any reason you dont want try-catch? it can help to have all the error handling code in catch and normal flow in try block. And you can't avoid the try-catch anyway because spawn may throw other errors

Reply:
Talked offline and mark as resolved. thanks.